src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs

```diff
@@ -0,0 +1,309 @@
+
+    /// netns
+    pub(crate) netns: Option<String>,
+
+    /// hypervisor config
+    pub(crate) config: HypervisorConfig,
+
+    /// vmm state
+    pub(crate) state: VmmState,
+
+    /// vmm instance
+    pub(crate) vmm_instance: VmmInstance,
+
+    /// hypervisor run dir
+    pub(crate) run_dir: String,
+
+    /// pending device
+    pub(crate) pending_devices: Vec<Device>,
+
+    /// cached block device
+    pub(crate) cached_block_devices: HashSet<String>,
+}
+
+impl DragonballInner {
+    pub fn new() -> DragonballInner {
+        DragonballInner {
+            id: "".to_string(),
+            vm_path: "".to_string(),
+            jailer_root: "".to_string(),
+            netns: None,
+            config: Default::default(),
+            pending_devices: vec![],
+            state: VmmState::NotReady,
+            jailed: false,
+            vmm_instance: VmmInstance::new(""),
+            run_dir: "".to_string(),
+            cached_block_devices: Default::default(),
+        }
+    }
+
+    pub(crate) async fn cold_start_vm(&mut self, timeout: i32) -> Result<()> {
+        info!(sl!(), "start sandbox cold");
+
+        self.set_vm_base_config().context("set vm base config")?;
+
+        // get rootfs driver
+        let rootfs_driver = self.config.blockdev_info.block_device_driver.clone();
+
+        // get kernel params
+        let mut kernel_params = KernelParams::new(self.config.debug_info.enable_debug);
+        kernel_params.append(&mut KernelParams::new_rootfs_kernel_params(&rootfs_driver));
+        kernel_params.append(&mut KernelParams::from_string(
+            &self.config.boot_info.kernel_params,
+        ));
+
+        // set boot source
+        let kernel_path = self.config.boot_info.kernel.clone();
+        self.set_boot_source(
+            &kernel_path,
+            &kernel_params
+                .to_string()
+                .context("kernel params to string")?,
+        )
+        .context("set_boot_source")?;
+
+        // get vm rootfs
+        let image = {
+            let initrd_path = self.config.boot_info.initrd.clone();
+            let image_path = self.config.boot_info.image.clone();
+            if !initrd_path.is_empty() {
+                Ok(initrd_path)
+            } else if !image_path.is_empty() {
+                Ok(image_path)
+            } else {
+                Err(anyhow!("failed to get image"))
+            }
+        }
+        .context("get image")?;
+        self.set_vm_rootfs(&image, &rootfs_driver)
+            .context("set vm rootfs")?;
+
+        // add pending devices
+        while let Some(dev) = self.pending_devices.pop() {
+            self.add_device(dev).await.context("add_device")?;
+        }
+
+        // start vmm and wait ready
+        self.start_vmm_instance().context("start vmm instance")?;
+        self.wait_vmm_ready(timeout).context("wait vmm")?;
+
+        Ok(())
+    }
+
+    pub(crate) fn run_vmm_server(&mut self, _timeout: i32) -> Result<()> {
```

Comment:
Unused parameter `_timeout`?

Reply:
Thanks for your review, already remove the parameter.